instrumentation/go.mongodb.org/mongo-driver/v2/mongo/otelmongo/mongo.go

```diff
@@ -0,0 +1,150 @@
+
+	hostname, port := peerInfo(evt)
+
+	attrs := []attribute.KeyValue{
+		semconv.DBSystemNameMongoDB,
+		semconv.DBOperationName(evt.CommandName),
+		semconv.DBNamespace(evt.DatabaseName),
+		semconv.NetworkPeerAddress(hostname),
+		semconv.NetworkPeerPort(port),
+		semconv.NetworkTransportTCP,
+	}
+	if !m.cfg.CommandAttributeDisabled {
+		attrs = append(attrs, semconv.DBQueryText(sanitizeCommand(evt.Command)))
+	}
+	if collection, err := extractCollection(evt); err == nil && collection != "" {
+		spanName = collection + "."
+		attrs = append(attrs, semconv.DBCollectionName(collection))
+	}
+	spanName += evt.CommandName
+	opts := []trace.SpanStartOption{
+		trace.WithSpanKind(trace.SpanKindClient),
+		trace.WithAttributes(attrs...),
+	}
+	_, span := m.cfg.Tracer.Start(ctx, spanName, opts...)
+	key := spanKey{
+		ConnectionID: evt.ConnectionID,
+		RequestID:    evt.RequestID,
+	}
+	m.Lock()
+	m.spans[key] = span
+	m.Unlock()
+}
+
+func (m *monitor) Succeeded(ctx context.Context, evt *event.CommandSucceededEvent) {
+	m.Finished(&evt.CommandFinishedEvent, nil)
+}
+
+func (m *monitor) Failed(ctx context.Context, evt *event.CommandFailedEvent) {
+	m.Finished(&evt.CommandFinishedEvent, fmt.Errorf("%w", evt.Failure))
+}
+
+func (m *monitor) Finished(evt *event.CommandFinishedEvent, err error) {
+	key := spanKey{
+		ConnectionID: evt.ConnectionID,
+		RequestID:    evt.RequestID,
+	}
+	m.Lock()
+	span, ok := m.spans[key]
+	if ok {
+		delete(m.spans, key)
+	}
+	m.Unlock()
+	if !ok {
+		return
+	}
+
+	if err != nil {
+		span.SetStatus(codes.Error, err.Error())
+	}
+
+	span.End()
+}
+
+// TODO sanitize values where possible, then re-enable `db.statement` span attributes default.
+// TODO limit maximum size.
+func sanitizeCommand(command bson.Raw) string {
+	b, _ := bson.MarshalExtJSON(command, false, false)
+	return string(b)
+}
+
+// extractCollection extracts the collection for the given mongodb command event.
+// For CRUD operations, this is the first key/value string pair in the bson
+// document where key == "<operation>" (e.g. key == "insert").
+// For database meta-level operations, such a key may not exist.
+func extractCollection(evt *event.CommandStartedEvent) (string, error) {
+	elt, err := evt.Command.IndexErr(0)
+	if err != nil {
+		return "", err
+	}
+	if key, err := elt.KeyErr(); err == nil && key == evt.CommandName {
+		var v bson.RawValue
+		if v, err = elt.ValueErr(); err != nil || v.Type != bson.TypeString {
+			return "", err
+		}
+		return v.StringValue(), nil
+	}
+	return "", errors.New("collection name not found")
+}
+
+// NewMonitor creates a new mongodb event CommandMonitor.
+func NewMonitor(opts ...Option) *event.CommandMonitor {
+	cfg := newConfig(opts...)
+	m := &monitor{
+		spans: make(map[spanKey]trace.Span),
+		cfg:   cfg,
+	}
+	return &event.CommandMonitor{
+		Started:   m.Started,
+		Succeeded: m.Succeeded,
+		Failed:    m.Failed,
+	}
+}
+
+func peerInfo(evt *event.CommandStartedEvent) (hostname string, port int) {
```

Comment:
Could we add tests for this method?

Reply:
I think you could construct a `*event.CommandStartedEvent` manually, like this, in unit tests.

```
	&event.CommandStartedEvent{
		ConnectionID: "foo",
	}
```

I want to confirm this logic can actually work, especially since this is a parsing logic written by hand. Do you consider using `net.SplitHostPort` lib?

Reply:
Since this was a copy/paste implementation of https://github.com/open-telemetry/opentelemetry-go-contrib/tree/323e373a6c15ae310bdd0617e3ed52d8cb8e4e6f/instrumentation/go.mongodb.org/mongo-driver/mongo/otelmongo, I hadn't considered it.  I'm open to checking out the feasibility of using `net.SplitHostPort`.

Reply:
@XSAM This is ready for review again.  I've updated the function to use `net.SplitHostPort`, and wrote some tests for it, as I discovered through the tests that there was pre-existing issues with the logic.  I also had to change the package name for `mongo_test.go` from `otelmongo_test` to `otelmongo` in order to access the private function for unit testing.